chamge/amqp/amqp-rpc.c

```diff
@@ -0,0 +1,700 @@
+        (char *) envelope.message.properties.content_type.bytes);
+  }
+
+  response_body =
+      g_strndup (envelope.message.body.bytes, envelope.message.body.len);
+  amqp_destroy_envelope (&envelope);
+
+  return response_body;
+}
+
+ChamgeAmqpRpc *
+chamge_amqp_rpc_new ()
+{
+  return CHAMGE_AMQP_RPC (g_object_new (CHAMGE_TYPE_AMQP_RPC, NULL));
+}
+
+static void
+chamge_amqp_rpc_dispose (GObject * object)
+{
+  ChamgeAmqpRpc *self = CHAMGE_AMQP_RPC (object);
+  ChamgeAmqpRpcPrivate *priv = chamge_amqp_rpc_get_instance_private (self);
+
+  if (priv->amqp_conn) {
+    /* close connection & destroy */
+    amqp_connection_close (priv->amqp_conn, AMQP_REPLY_SUCCESS);
+    amqp_destroy_connection (priv->amqp_conn);
+    priv->amqp_conn = NULL;
+  }
+
+  G_OBJECT_CLASS (chamge_amqp_rpc_parent_class)->dispose (object);
+}
+
+static void
+chamge_amqp_rpc_finalize (GObject * object)
+{
+  ChamgeAmqpRpc *self = CHAMGE_AMQP_RPC (object);
+  ChamgeAmqpRpcPrivate *priv = chamge_amqp_rpc_get_instance_private (self);
+
+  G_OBJECT_CLASS (chamge_amqp_rpc_parent_class)->finalize (object);
+}
+
+static void
+chamge_amqp_rpc_get_property (GObject * object,
+    guint prop_id, GValue * value, GParamSpec * pspec)
+{
+  ChamgeAmqpRpc *self = CHAMGE_AMQP_RPC (object);
+  ChamgeAmqpRpcPrivate *priv = chamge_amqp_rpc_get_instance_private (self);
+
+  switch ((_ChamgeAmqpRpcProperty) prop_id) {
+    case PROP_HOSTNAME:
+      g_value_set_string (value, priv->connection_info.host);
+      break;
+    case PROP_PORT:
+      g_value_set_uint (value, priv->connection_info.port);
+      break;
+    case PROP_CHANNEL:
+      g_value_set_uint (value, priv->channel);
+      break;
+    case PROP_VHOST:
+      g_value_set_string (value, priv->connection_info.vhost);
+      break;
+    case PROP_ID:
+      g_value_set_string (value, priv->connection_info.user);
+      break;
+    case PROP_PWD:
+      g_value_set_string (value, priv->connection_info.password);
+      break;
+    case PROP_URL:
+      g_value_set_string (value, priv->amqp_url);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+static void
+_amqp_parse_url (const gchar * url, rpc_connection_info * const conn_info)
+{
+  g_autofree gchar *s = g_strdup (url);
+  amqp_parse_url (url, conn_info);
+  g_debug ("url parsed : host[%s] vhost[%s]", conn_info->host,
+      conn_info->vhost);
+  g_debug ("    user[%s] pwd[%s] port[%d] ssl[%d]", (conn_info)->user,
+      (conn_info)->password, (conn_info)->port);
+}
+
+static void
+chamge_amqp_rpc_set_property (GObject * object,
+    guint prop_id, const GValue * value, GParamSpec * pspec)
+{
+  ChamgeAmqpRpc *self = CHAMGE_AMQP_RPC (object);
+  ChamgeAmqpRpcPrivate *priv = chamge_amqp_rpc_get_instance_private (self);
+
+  switch ((_ChamgeAmqpRpcProperty) prop_id) {
+    case PROP_HOSTNAME:
+      priv->connection_info.host = g_value_dup_string (value);
```

Comment:
this way will cause memory leakage if user sets more than once. 
`g_free()` should be called before copying string.